rfcs/20221020-tensor-indexed-updates.md

```diff
@@ -0,0 +1,468 @@
+so the overall syntax might look like:
+```
+x = tf.copy_and_update(x, np.index_exp[:, :, 2], 0)
+```
+
+- Advantages
+  - More explicit that the tensor is being copied.
+  - More consistent with exising TensorFlow free-function syntax.
+
+- Disadvantages
+  - More verbose.
+  - `np.index_exp` is not very widely known or used.
+  - New untested API.
+  - Lack of consistency with JAX.
+  
+In discussions with the JAX team, they initially started with a syntax like this
+but then eventually switched to the current `.at` syntax due to user feedback.
+
+#### Augment Existing Tensor Scatter Methods
+
+We do already have a suite of ops that copy-and-modify the tensor:
+`tf.tensor_scatter_nd_*`.  These accept single-index and integer-tensor
+indexing. We could add new functionality to them to allow striding and boolean
+masks.  However, there are a few drawbacks:
+- The scatter/gather operations currently have a very specific meaning,
+  and have a 1-1 correspondence with dedicated ops/kernels
+  (e.g. `TensorScatterUpdate`).  If we change the meaning this will no longer
+  apply, and gather/scatter would act more like dispatchers.
+- The shape of index arguments to gather/scatter are inconsistent with the Numpy
+  convention: one is the transpose of the other.  For example,
+  ```python
+  x = np.reshape(np.arange(6), [2, 3])
+  indices = tuple([[0, 1], [0, 2]])
+  
+  y = x[indices]                               # [0, 5]
+  z1 = tf.gather_nd(x, indices)                # [1, 2]
+  z2 = tf.gather_nd(x, tf.transpose(indices))  # [0, 5]
+  ```
+  
+We would therefore either need to break consistency with NumPy, or break
+backward compatibility with existing usages - neither of which is desirable.
+
+#### Mutable Tensors or tf.Variables
+
+Making existing tensors mutable is not possible nor desirable due to
+TensorFlow's current design.  We considered introducing a `MutableTensor` object
+that could modify and track an underlying tensor object.
+
+However, this would be very similar to `tf.Variable`, which does already allow
+slice assignments via the `.assign()` member function.  Instead we propose to
+allow direct assignment via `tf.Variable.__setitem__` to update a variable
+in-place.  Technically `tf.Variable` could then be used as a proxy to update a
+`tf.Tensor`:
+```python
+var_x = tf.Variable(x)
+var_x[:,:,2] = 0             # Modify the Variable's underlying contents.
+x = var_x.read_value()       # Extract the modified contents.
+```
+Forcing this to be the main update syntax for regular tensors seems a bit
+clunky and heavy-handed, since we don't need all that `tf.Variable` provides.
+The `.at[]` syntax provides a more direct option.
+  
+### Performance Implications
+
+No new ops are to be added (at least not initially).  Instead, we propose
+creating a simplified API around existing ones.  Performance for existing ops
+will be unaffected.
+
+Such a copy-and-update method can lead to poor practices: users may overuse and
```

Comment:
Is there the risk that users start to update single elements in a loop with this new syntax?

Reply:
Yes, which would lead to excessive copying of tensors.

As mentioned in this paragraph, we can try to emphasize in the docs that this shouldn't be done, and it's not something we've observed with `scatter_nd_update`, which could also be misused in the same way.

Reply:
Yes I saw bu I think it's a bit less natural "to abuse" `scatter_nd_update` in a loop instead of a numpy like syntax.
Of course it will be slow also on numpy as we suppose to prefer a "vectorized" use of the op so generally this kind of "individual" access could probably rely on ufunc (See https://github.com/google/mlir-npcomp/issues/1). But in our case we need also to add the extra copy overhead.

Also I don't know if this misuse could be detected on autograph tracing at least to emit a warning (like the one we have for excessive retracing).

Reply:
> Also I don't know if this misuse could be detected on autograph tracing at least to emit a warning

Potentially.  I'm sure it's worth it until if/when we see that this kind of abuse is common.  For those who do care about performance, benchmarking/profiling the model will likely expose these repeated calls as the issue.

Reply:
Do you have any feedback instead about the mentioned "ufunc like" needs?

Reply:
I'm not sure I understand what the "ufunc like" needs are.  I agree if a user is modifying a tensor in a loop one element at a time, they could use a ufunc.  The `tensor.at[index_expression].apply(ufunc)` should support this.

Reply:
Sure but I meant how this is mapped efficiently "down to the stack"? I meant if we suppose to use `ufunc` when we need to access to the Tensor in without the compositional TF/*HLO ops can we have something like e.g. numba [@vectorize decorators](https://numba.pydata.org/numba-doc/dev/user/vectorize.html) or probably similar `jax.numpy.vectorize`?

Reply:
Like [tf.map_fn](https://www.tensorflow.org/api_docs/python/tf/map_fn)?  I agree an efficient vectorized element-wise operation would be useful.  This is not what we're proposing here though.  Well, I suppose `.at[...].apply(ufunc)` should probably do something smart like that.  The focus here is on the update API though, so we'll just re-use whatever already exists in TF for the implementation details to start.  These can be improved later via other additions like a `tf.vectorize` API.

Reply:
>  Well, I suppose .at[...].apply(ufunc) should probably do something smart like that.

Yes, thank you for the clarification. I supposed that this was the right occasion but if we'll just re-use whatever already exists it is still not the case.

Reply:
It would sill nice if we could put this in a notice as a future work or just to disambiguate the perimeter of the RFC.